### ISSUE TITLE
CUDA status Error: file: ..\..\src\dark_cuda.c : cuda_push_array() 

### ISSUE THREAD

**Issue by Kmarconi** [AlexeyAB/darknet#3392]
Hi, first, thank you for your amazing GitHub tutorial for using darknet in any case.

It's been a few days since I'm working on a project which need to use 3 rtsp camera and count certain type of object in each in real time. I'm using yolov3 weight and cfg with batch and subdivisions to 1 for testing and when I'm using only 2 camera it seems to run pretty well but if i plug the third one, I will Always get the eroor in the title of the issue. I've been googling it a lot but I haven't been able to find Something which could help me. I'm working on a Quadro P2000 with 32GB RAM and Xeon CPU.

Would be amazing if you could help me to find a way to resolve this error. 

Best regards ! 

Kévin Marconi

PS : French student so sorry if my English is not that good !

**Comment by AlexeyAB** [comment]
@Kmarconi Hi,

There is no enough GPU-VRAM for 3 yolo v3 instances.
* Or use GPU with higher VRAM
* Or implement by yourself application (Python/C++) that processes 3 rtsp streams by using 1 yolo instance

**Comment by Kmarconi** [comment]
Thanks for you quick answer,

I'm launching them as thread ( so 3 thread) but by one personnal python script which load the yolo weight in the "init" part of it. I've tried to make 3 different instances of my detection script and effectively, it was loading the weight 3 time and it was crashing all the time. But here i'm only loading it once at the beginning of it so …. I might be wrong but I Don't understand

**Comment by Kmarconi** [comment]
I also have this error when the dark_cuda error is not their : CUDA Error Prev: unspecified launch Failure 
Don't know if it can guide you

**Comment by keko950** [comment]
I'm using darknet in a similar case and im not having any trouble, also using python too!
You are probably freeing/deleting an image before detections are made on it, its more like a memory leak.

If you post your python code we can check it out!

**Comment by Kmarconi** [comment]
The problem is that i'm working for a society and my code should remain secret … Can I see yours to compare ? What I'm doing is that i'm only taking a frame of the video Stream every 10 frame to not do too much operation, i'm saving this frame as a temp image to send it to darknet, and if darknet return what I want, I increase the counting number, otherwise i just pass and loop other this process

**Comment by keko950** [comment]
I cant share my code, sorry.

So, im assuming you are using a modified version of darknet.py which remains in this repo, right? And you said your program only segfaults when you use 3 cameras right? Are you checking before send the image to darknet if the image is well formed? Sometimes rtsp streams can send you an incorrect image, check image shapes, check also if your images are in rgb

**Comment by Kmarconi** [comment]
I just tested with 3 Capture based on video files and not rtsp camera and the same error appeared : CUDA Error Prev: unspecified launch Failure … Even with only 2 video files running but the error appear later, like if the memory was never getting free and it was getting filled all the time. 
I'm using the darknet.py file with a modified version of performDetect to do what i want to do about counting.

I heard about the fact that rtsp streams can send incorrect image but the error also appear with video file and usb webcam so ....

**Comment by keko950** [comment]
Run in another terminal:
$ watch nvidia-smi
and then run your script, check if memory increases.

Are you using the function array_to_image(arr)? 

`def array_to_image(arr):


    import numpy as np
    # need to return old values to avoid python freeing memory
    arr = arr.transpose(2,0,1)
    c = arr.shape[0]
    h = arr.shape[1]
    w = arr.shape[2]
    arr = np.ascontiguousarray(arr.flat, dtype=np.float32) / 255.0
    data = arr.ctypes.data_as(POINTER(c_float))
    im = IMAGE(w,h,c,data)
    return im, arr`

**Comment by Kmarconi** [comment]
No I'm not, it is actually commented in my detect_image function

**Comment by keko950** [comment]
To each frame you want to detect you should convert to image first:

` 
               im, arr = array_to_image(f)


               detects = self._detect(
                    self.network,
                    self.meta,
                    im,
                    ..
                )`

**Comment by Kmarconi** [comment]
Sorry, I Don't know if you have time for it, it is already very cool to help me through those CUDA error, but can you explain a little bit more if possible ? I need to modify the detect_image function and the detect too ? I saw that i need to comment " free_image(im)" and uncomment the section : 
`import cv2
    custom_image_bgr = cv2.imread(image) # use: detect(,,imagePath,)
    custom_image = cv2.cvtColor(custom_image_bgr, cv2.COLOR_BGR2RGB)
    custom_image = cv2.resize(custom_image,(lib.network_width(net), lib.network_height(net)), interpolation = cv2.INTER_LINEAR)
    import scipy.misc
    custom_image = scipy.misc.imread(image)
    im, arr = array_to_image(custom_image)`

Is it good ?

**Comment by keko950** [comment]
im assuming you are using opencv to read frames from the stream, by default opencv works with BGR image color, so you have to change it to RGB with custom_image = cv2.cvtColor(custom_image_bgr, cv2.COLOR_BGR2RGB).

resize is not necessary since darknet does it himself inside (as far as i know, correct me if im wrong @AlexeyAB ) so you can comment those lines.

you have to comment free_image, right, you dont need it.

this last part "im, arr = array_to_image(custom_image)" is important, i had some errors weeks ago caused by not using that function.

So, basically yo have to do is:

getImageFromStream()
ConvertToRgb()
array_to_image()
detection()

**Comment by Kmarconi** [comment]
Actually what my detect and detect_image ( the beginning) are looking like, just tested with 1 video and it is working, will see if I get the error with 2 or 3 cameras.

Will keep you update ! Thank you 
`

def detect(net, meta, image, thresh=.5, hier_thresh=.5, nms=.45, debug= False):
    """
    Performs the meat of the detection
    """
    #pylint: disable= C0321
    #im = load_image(image, 0, 0)
    #if debug: print("Loaded image")
    ret = detect_image(net, meta, image, thresh, hier_thresh, nms, debug)
    #free_image(im)
    if debug: print("freed image")
    return ret

def detect_image(net, meta, im, thresh=.5, hier_thresh=.5, nms=.45, debug= False):
    import cv2
    custom_image_bgr = cv2.imread(im) # use: detect(,,imagePath,)
    custom_image = cv2.cvtColor(custom_image_bgr, cv2.COLOR_BGR2RGB)
    custom_image = cv2.resize(custom_image,(lib.network_width(net), lib.network_height(net)), interpolation = cv2.INTER_LINEAR)
    import scipy.misc
    custom_image = scipy.misc.imread(im)
    im, arr = array_to_image(custom_image)`

**Comment by Kmarconi** [comment]
Ok I will comment the resize line :) So I'm not getting any more the launch Failure error but now i'm getting this one : CUDA Error Prev: misaligned address , never saw it before, did you got it too in your experiment ?

**Comment by Kmarconi** [comment]
Nevermind, I just got the CUDA Error Prev: unspecified launch Failure while running my script on 3 videos ...

**Comment by keko950** [comment]
Hummmm... its kinda strange, i didnt ever receive that error. Can you create an independent script and reproduce the error so you can send it to me?

**Comment by Kmarconi** [comment]
I had to left work for today but just before I left, i've launched my script three time on the same rtsp videostream and it worked the whole time without any error so it might be because i was processing on video file. I also observed that cv2.VideoCapture was taking a lot more of cpu with video file than with rtsp, might be related, don't know. Will keep you updated tomorrow! Thank you, have a nice day :)

**Comment by keko950** [comment]
No problem, just tell me if its solved! :)

**Comment by Kmarconi** [comment]
Hi :) So i was testing with 3 rtsp cameras and it was all going well but as i'm writing this message, I just got CUDA status Error: file: ..\..\src\dark_cuda.c : cuda_push_array() , CUDA Error Prev: unspecified launch Failure     for another time …. I'm doing the process you told me to do : 
getImageFromStream()
ConvertToRgb()
array_to_image()
detection()

but still ^^' Verry strange >< My nvidia-smi is always like this, only WDDM is moving 

`
+-----------------------------------------------------------------------------+
| NVIDIA-SMI 430.64       Driver Version: 430.64       CUDA Version: 10.2     |
|-------------------------------+----------------------+----------------------+
| GPU  Name            TCC/WDDM | Bus-Id        Disp.A | Volatile Uncorr. ECC |
| Fan  Temp  Perf  Pwr:Usage/Cap|         Memory-Usage | GPU-Util  Compute M. |
|===============================+======================+======================|
|   0  Quadro P2000       WDDM  | 00000000:01:00.0  On |                  N/A |
| 56%   55C    P0    52W /  75W |   2030MiB /  5120MiB |     39%      Default |
+-------------------------------+----------------------+----------------------+

+-----------------------------------------------------------------------------+
| Processes:                                                       GPU Memory |
|  GPU       PID   Type   Process name                             Usage      |
|=============================================================================|
|    0      1320    C+G   Insufficient Permissions                   N/A      |
|    0      4508    C+G   ...mmersiveControlPanel\SystemSettings.exe N/A      |
|    0      6360    C+G   ...osoft.LockApp_cw5n1h2txyewy\LockApp.exe N/A      |
|    0      7544    C+G   C:\Windows\explorer.exe                    N/A      |
|    0      8360    C+G   ...t_cw5n1h2txyewy\ShellExperienceHost.exe N/A      |
|    0      8708    C+G   ...dows.Cortana_cw5n1h2txyewy\SearchUI.exe N/A      |
|    0      9488    C+G   ...oftEdge_8wekyb3d8bbwe\MicrosoftEdge.exe N/A      |
|    0     10408    C+G   ...19011.0_x64__8wekyb3d8bbwe\Video.UI.exe N/A      |
|    0     11784      C   ...ta\Local\Continuum\anaconda3\python.exe N/A      |
+-----------------------------------------------------------------------------+
`
The VRAM never gets beyond 2300 MB so i don't understand why I'm getting the unspecified launch Failure error. About reproducing a script to give it to you I'm not sure how I can do it without revealing the process of it but I will try to describ it as the most complete I can do.

So first, i'm launching a script which load the yolov3 weight and let me choose which camera on my network I want to use. I select the ones i want, than i'm starting a detection thread on each of them which is looping while cap.isOpened() and starting a detection every 10 frame so that it is not overloaded. To detect, i'm calling the performDetect() function and if the object i want to count is detected It is returning True, otherwise, false. 


I hope it is clear, tell me if its not :) 


Thank you and have a good day

**Comment by keko950** [comment]
Oh, so you are trying to inference in three isolated threads? I think thats the problem, you cant do that!
Seems like your threads are going into a race condition in some value which makes cuda crash. 
If you want to make three independent threads you should use multi gpu, and running three instances of the detector i guess.

My advice to you, is just simple create a frame queue for each stream, and in the main thread run detection SEQUENTALLY.

So your script shall remain:

Thread1 -> GettingFrames -> AppendToFrameQueue1
Thread2 -> GettingFrames -> AppendToFrameQueue2
Thread3 -> GettingFrames -> AppendToFrameQueue3

MainThread: create a infinite loop that will get frames from that queues and process them, get detections, work with the returned data!

**Comment by Kmarconi** [comment]
Really interesting way of doing it, never thought about it :o Thanks for the advice, tbh i've started to really work hard with opencv and cuda only 2 months ago so I hope I will be able to do it because it seems to be a really a good way to have good performance. 

Thanks !

**Comment by keko950** [comment]
You should always have an independent thread for decoding frames from streams, so the thread will not remain blocked doing process work, so, if you append each decoded frame to a queue which is visible for the main thread, you can make detection on that frames meanwhile the decoder thread will keep decoding more frames, :)
Also, check this!
https://www.pyimagesearch.com/2017/02/06/faster-video-file-fps-with-cv2-videocapture-and-opencv/

Something like this
![Captura de pantalla de 2019-06-13 10-17-08](https://user-images.githubusercontent.com/9590497/59415792-72c30280-8dc4-11e9-980f-77e0a00d9120.png)

**Comment by Kmarconi** [comment]
Ok , will try it, thanks :)

**Comment by Kmarconi** [comment]
Might be stupid question ( even if there is no stupid question i know x) ) but i did my decoding _rtsp thread where i'm reading the Stream, and stocking the frame every 10 frame in a list like this : 

`def run(self):

        #self.vs

        i=0

        while True:

            frame = self.vs.read()

            frame = imutils.resize(frame, width=400)

            if i==10:

                self.tabframe.append(frame)

                print(len(self.tabframe))

                #print(i)

                i=0

            i+=1

    def getFrame(self):

        try:

            print(type(self.tabframe[0]))

            return self.tabframe[0]

            #self.tabframe.pop(0)

        except:

            pass#print("No frame to work with")`


and when in my main, i'm calling the getframe() function from my thread, it is returning a NoneType whereas the thing is the list of my thread is an nd.array ><

**Comment by keko950** [comment]
Absolutely wrong, you must only call join to stop your thread and wait to resume his tasks.
Check that streaming is openingh right, also remember that opening a stream takes a while, so maybe you are trying to get the frame too early. Your code seems okay by the way

**Comment by Kmarconi** [comment]
So, I finally been able to convert my ex script to the new way and for now, I've tested with 1 camera and the counting is working but like 10 or even 15 second after it should be. Did you encountered this problem ?

**Comment by keko950** [comment]
What problem? What you mean by counting is working?

**Comment by Kmarconi** [comment]
(updated my text)

**Comment by keko950** [comment]
You are appending a frame each 10 frames in 3 streams, thats 3frames per second, and youre popping too slow so it will remain as a leak, try not using lists or arrays. Use a circular queue instead with a max size https://www.pythoncentral.io/circular-queue/

**Comment by Kmarconi** [comment]
Is it ok to talk by mail to not spam this discussion ?

**Comment by Kmarconi** [comment]
Hi ! Fresh brain after sleeping is better for programming x) I found the problem of the queue thing. With the WebcamVideoStream from imutils Library, there isn't the ret,frame return but just frame ( or i haven't seen it, if so, tell me ^^ ). So i just tried with classic cv2.VideoCapture and my queue never get more element than 1 in it so that's great ! Will keep you updated

**Comment by Kmarconi** [comment]
Update : Just saw that i can modifie the webcamvideostream script in imutils package to add return the "Grabbed" variable, will try with it

**Comment by Kmarconi** [comment]
Working like a charm with 3 rtsp camera, CPU only at 18%, 1.8GB of RAM used , pretty cool !

**Comment by keko950** [comment]
Sorry, i was out this weekend, im glad it worked! :)

Have a nice day

**Comment by xinsuinizhuan** [comment]
Your problem have solved? I meet the same problem.

**Comment by xinsuinizhuan** [comment]
I only create three nets at same time，and dit not detecte! When i run it the forth or fiveth, it crash!

**Comment by Kmarconi** [comment]
Yeah read the thread above :) You canno't do 3 nets at the same time, you have to do only one net and queue your videos and loop over your queues with darknet

**Comment by LionelLeee** [comment]
> 搭配3台rtsp相机，就像魅力一样工作，CPU仅占18％，使用1.8GB RAM，非常酷！

Does the implementation of this approach affect the speed of each instance? What is the impact on fps?

**Comment by LionelLeee** [comment]
> Working like a charm with 3 rtsp camera, CPU only at 18%, 1.8GB of RAM used , pretty cool !

Does the implementation of this approach affect the speed of each instance? What is the impact on fps?

**Comment by LionelLeee** [comment]
@Kmarconi Take the above method if you capture multiple video sources. Then the time for processing a single video will increase, which is not suitable for real-time display.

**Comment by mayankcstech** [comment]
@Kmarconi after your problem was solved, what was your CUDA utilization. Are you able to utilize it fully (~99%)?

**Comment by Kmarconi** [comment]
Sorry for commenting late @LionelLeee, I was working with a quadro p2000 and I've been able to work with 3 rtsp cameras in real time and haven't seen any effect on fps since the inference with darknet is really fast.

@mayankcstech If I remember well, I wasn't using it fully at all since I didn't needed my program to inference each of the camera's frames. But if I remember well, I tried to push it to the limit by doing inference on each frames and yeah I was always above 90% and even 99%
